cosmpy/aerial/contract/__init__.py

```diff
@@ -0,0 +1,250 @@
+# -*- coding: utf-8 -*-
+# ------------------------------------------------------------------------------
+#
+#   Copyright 2018-2021 Fetch.AI Limited
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# ------------------------------------------------------------------------------
+
+import json
+from datetime import datetime
+from typing import Any, Optional
+
+from cosmpy.aerial import LedgerClient
+from cosmpy.aerial.contract.cosmwasm import (
+    create_cosmwasm_execute_msg,
+    create_cosmwasm_instantiate_msg,
+    create_cosmwasm_store_code_msg,
+)
+from cosmpy.aerial.tx import SigningCfg, Transaction
+from cosmpy.aerial.tx_helpers import SubmittedTx
+from cosmpy.crypto.address import Address
+from cosmpy.crypto.hashfuncs import sha256
+from cosmpy.crypto.keypairs import PrivateKey
+from cosmpy.protos.cosmos.base.query.v1beta1.pagination_pb2 import PageRequest
+from cosmpy.protos.cosmwasm.wasm.v1.query_pb2 import (
+    QueryCodesRequest,
+    QuerySmartContractStateRequest,
+)
+
+
+def _compute_digest(path: str) -> bytes:
+    with open(path, "rb") as input_file:
+        return sha256(input_file.read())
+
+
+def _generate_label(digest: bytes) -> str:
+    now = datetime.utcnow()
+    return f"{digest.hex()[:14]}-{now.strftime('%Y%m%d%H%M%S')}"
+
+
+class LedgerContract:
+    def __init__(
+        self, path: str, client: LedgerClient, address: Optional[Address] = None
+    ):
+        self._path = path
```

Comment:
We might be able to pass the path as Optional and then we should be able to determine if the same .wasm binary already exists on-chain so we can retrieve code_id and then instantiate new contract_address or just connect to exiting already instantiated contract_address.

Ed, correct me if I am wrong, but it looks like you want an instance of Contract to represent actual 1 deployed contract_address (And have contract_address stored in member variable) rather than having to provide contract_address on each call.

There are 3 possible scenarios
1. Store contract, instantiate -> get contract_address (Provide contract .wasm path in constructor)
2. Just instantiate -> get contract_address (Provide code_id in constructor)
3. Provide contract_address in constructor

Reply:
Often we just want to connect to an already deployed contract without having its binary.